File: n_puzzle/heuristics.py
from n_puzzle.puzzle import Puzzle


target_indexes = []
target_map = []


def hemming(puzzle: Puzzle) -> int:
    out = 0
    for i, elem in enumerate(puzzle.board):
        out += int(elem != target_map[i])
    return out


def get_manhattan_score(side_len: int, elem: int, i: int) -> int:
    current_y = i // side_len
    current_x = i % side_len
    goal_y, goal_x = target_indexes[elem]
    return abs(goal_x - current_x) + abs(goal_y - current_y)


def manhattan(puzzle: Puzzle) -> int:
    out = 0
    for i, elem in enumerate(puzzle.board):
        out += get_manhattan_score(puzzle.side_len, elem, i)
    return out


def phased_manh(puzzle: Puzzle) -> int:
    out = 0

    for i in range(puzzle.puzzle_len):
        if puzzle.board[i] < puzzle.side_len * (puzzle.side_len - 2):
            out += get_manhattan_score(puzzle.side_len, puzzle.board[i], i) * 4
        elif puzzle.board[i] % puzzle.side_len < puzzle.side_len - 2:
            out += get_manhattan_score(puzzle.side_len, puzzle.board[i], i) * 2
        else:
            out += get_manhattan_score(puzzle.side_len, puzzle.board[i], i)
    return out


def init_weights(weights: list, puzzle: Puzzle):

    for i in range(puzzle.puzzle_len - 1, -1, -1):
        if i / puzzle.side_len < puzzle.side_len - 2:
            weights[i] = (puzzle.side_len - int(i / puzzle.side_len)) * 2
        elif i % puzzle.side_len < puzzle.side_len - 2:
            weights[i] = puzzle.side_len - int(i / puzzle.side_len)
            weights[i] += puzzle.side_len - i % puzzle.side_len


def rowwise_manhattan(puzzle: Puzzle):

    out = 0
    weights = [1 for i in range(puzzle.puzzle_len)]

    if not weights:
        init_weights(weights, puzzle)
    for i in range(puzzle.puzzle_len):
        out += get_manhattan_score(puzzle.side_len, puzzle.board[i], i) * weights[puzzle.board[i]]
    return out


def uniform(puzzle: Puzzle) -> int:

    for i in range(puzzle.puzzle_len):
        if puzzle.board[i] != target_map[i]:
            return 1
    return 0


def greedy(puzzle: Puzzle):

    puzzle.g = 0
    return phased_manh(puzzle)

File: cpp/heuristics.cpp
#include "n_puzzle.hpp"

int	Puzzle::side_len;

int	hemming(const Puzzle *puzzle, const puzzle_config_t& goal)
{
	int out = 0;
	for (int i = 0; i < static_cast<int>(puzzle->map.size()); ++i)
		out += (i != goal[puzzle->map[i]]);
	return (out);
}

inline static int get_manhattan_score(int i, int j, int side_len)
/*
	Returns the number of steps to move from position i to position j.
*/
{
	return abs(i / side_len - j / side_len) + abs(i % side_len - j % side_len);
}

int	manhattan(const Puzzle *puzzle, const puzzle_config_t& goal)
{
	int out = 0;
	int size = puzzle->map.size();

	for (int i = 0; i < size; ++i)
		out += get_manhattan_score(i, goal[puzzle->map[i]], Puzzle::side_len);
	return (out);
}

int	phased_manhattan(const Puzzle *puzzle, const puzzle_config_t& goal)
/*
	This heuristic is only intended for a classic n-puzzle.
*/
{
	int out = 0;

	for (int i = 0; i < static_cast<int>(puzzle->map.size()); ++i)
	{
		if (puzzle->map[i] < Puzzle::side_len * (Puzzle::side_len - 2))
			out += get_manhattan_score(goal[puzzle->map[i]], i, Puzzle::side_len) * 4;
		else if (puzzle->map[i] % Puzzle::side_len < Puzzle::side_len - 2)
			out += get_manhattan_score(goal[puzzle->map[i]], i, Puzzle::side_len) * 2;
		else
			out += get_manhattan_score(goal[puzzle->map[i]], i, Puzzle::side_len);
	}
	return out;
}

static void init_weights(std::vector<int>& weights)
{
	weights.resize(Puzzle::side_len * Puzzle::side_len, 1);
	for (int i = weights.size() - 1; i >= 0; --i)
	{
		if (i / Puzzle::side_len < Puzzle::side_len - 2)
		{
			weights[i] = (Puzzle::side_len - i / Puzzle::side_len) * 2;
		}
		else if (i % Puzzle::side_len < Puzzle::side_len - 2)
		{
			weights[i] = (Puzzle::side_len - i / Puzzle::side_len);
			weights[i] += Puzzle::side_len - i % Puzzle::side_len;
		}
	}
}

int	rowwise_manhattan(const Puzzle *puzzle, const puzzle_config_t& goal)
/*
	This heuristic is only intended for a classic n-puzzle.
*/
{
	int out = 0;
	static std::vector<int> weights;

	if (!weights.size())
		init_weights(weights);
	for (int i = 0; i < static_cast<int>(puzzle->map.size()); ++i)
		out += get_manhattan_score(goal[puzzle->map[i]], i, Puzzle::side_len) * weights[puzzle->map[i]];
	return out;
}

int	uniform(const Puzzle *puzzle, const puzzle_config_t& goal)
{
	for (int i = 0; i < static_cast<int>(puzzle->map.size()); ++i)
		if (goal[puzzle->map[i]] != i)
			return (1);
	return (0);
}

int	greedy_manhattan(const Puzzle *puzzle, const puzzle_config_t& goal)
{
	//hacky-hacky.
	uintptr_t addr = reinterpret_cast<uintptr_t>(puzzle);
	Puzzle* deconst = reinterpret_cast<Puzzle*>(addr);
	deconst->g = 0;
	return manhattan(puzzle, goal);
}
